### PR TITLE
Add additional debugging metadata using the SourceDebugExtension attribute

### PR DESCRIPTION
**Pull request by Minecrell**
**TL;DR:** With these changes it is possible to debug Mixins using their original source files, the same way you can debug regular Java classes. It requires a modification of the IDE - for IntelliJ IDEA I've made these changes in a PR to [MinecraftDev](https://minecraftdev.org): https://github.com/minecraft-dev/MinecraftDev/pull/115. Once all changes are merged installing MinecraftDev is the only step needed to debug Mixins.

---

### The issue
Mixin classes are merged into the target classes at runtime. Because of that, debugging Mixins usually results in jumping around on invalid lines in the source of the target class instead of in the actual Mixin code.

The debugger resolves the source location using the class name and the `SourceFile` attribute of the Java class file, which is limited to a single source file. Consequently, debugging methods in the target class and a Mixin is not possible at the same time: the `SourceFile` can be only set to  the original source file or the source file of the Mixin.

### Using decompiled code
Up to now, the only solution for this problem was to use special support in the IDE (primarily IntelliJ IDEA) to debug using decompiled classes instead of the original sources. However, while you are able to see the classes exactly the way they are executed at runtime (including the invocation of injectors etc) it has several disadvantages:

- Requires manual setup to configure Mixin to dump the bytecode and add it to the module classpath
- Comments and the original code formatting is removed due to the decompilation
- Breakpoints can be only set after the Mixin has already been applied once. To add a breakpoint for a new Mixin you have to run the application once to dump the new bytecode and let the IDE refresh all classes until you can actually see your changes in the target class.

### Source debug extension
As already mentioned, Java supports only a single `SourceFile` attribute in the class file. This is not only a problem for Mixin, but also for other projects, primarily programming languages which are compiled to Java source code before getting compiled to the final bytecode.

A solution for this problem was added to Java as part of [JSR-045 (Debugging Support for Other Languages)](https://jcp.org/en/jsr/detail?id=45). It describes a new attribute in the Java class file format, the `SourceDebugExtension` attribute, together with a standardized Source Map Format (SMAP).

SMAP was primarily developed for languages which are first compiled to Java source files. It can map the final line numbers in the bytecode to other line numbers in one or multiple source files.

On the other hand, Mixin keeps all original line numbers from the Mixins that are applied to the target class. Therefore the line numbers in the final class are not unique, and cannot be mapped reliably to the original source.

### The solution
A solution would be to assign unique line numbers to all merged Mixin methods. However, this would make reading stack traces even harder than it already is. Right now you can use the line number to find the Mixin which is causing an exception.

A much easier solution is to encode a custom format in the `SourceDebugExtension` which is specific to Mixin. The [Java class file format specification](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.11) does not dictate a specific format for the attribute - it is possible to store any string in it. All tools I've checked verify if the string is a SMAP (using the `SMAP` prefix) before parsing it. It is also [exposed in raw form in Java's debugging interface (JDI)](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/ReferenceType.html#sourceDebugExtension--), so we can parse it in the debugger to provide Mixin debugging.

In general, the only thing the debugger needs to know is the Mixin the method in the target class was merged from. I've used a very simple format to encode the metadata in the attribute:

```
MIXIN
com.example.mixin.MixinExample run()V handler$onRun$zza000(Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfo;)V
com.example.mixin.MixinExtra runExtra()V
```

The `MIXIN` prefix allows us to detect encoded data as Mixin source debug extension - this is very similar to the solution in the SMAP format which uses the `SMAP` indentifier.

### The implementation
I've added a new Mixin option `-Dmixin.sourceDebugExtension` which can be used to enable the generation of the attribute. I decided to disable it by default to prevent potential problems with other tools that are using the same attribute. In most cases there is no need to generate it because it is only usable by extending the IDE debugger to parse the additional metadata.

The metadata is generated by parsing all `@MixinMerged` annotations after all Mixins have been applied. I'm quite sure there are better ways to implement it, but that seemed like the easiest solution to me.

### IDE implementation
I've implemented debugging support for the IntelliJ IDEA plugin [MinecraftDev](https://github.com/minecraft-dev/MinecraftDev). It parses the additional metadata and uses it to jump to the correct source of the method. It will also redirect breakpoints to the target class so it will properly pause the execution at runtime.

See https://github.com/minecraft-dev/MinecraftDev/pull/115 for some screenshots of the Mixin debugging.

### CODE REVIEW

**Comment by Mumfrey** [comment]
1. Why is this a PR and not an issue? Open an issue **first** to discuss things with me before making PRs out of the blue.
2. I don't see why you wouldn't use the SMAP format for this, there is already code to support it.
3. Calling the option `sourceDebugExtension` is misleading because it doesn't actually enable the standard SMAP format, it should be called something which references the fact that it only supports one particular, InteliJ plugin.

Honestly I think if you want to do this, it would be more appropriate to support it by providing a mechanism to add extra plugins into Mixin, and then implement the IntelliJ plugin on that.

Seriously, this should have been an issue. I'd appreciate you closing it and opening discussion in the proper manner.
